### ISSUE TITLE
Idea for removing the static classes

### ISSUE THREAD

**Issue by designermonkey** [symphonycms/symphony-2#2458]
I just want to leave this here: https://github.com/rdlowrey/Auryn

One of my biggest wishes for v3.0.0 is to finally remove all the static classes. I've played with the idea loads, found lots of methods to do it, and none of them are easy to do.

I found this IoC pattern the other day, and it sounds like the right way to go. All the classes we use as statics at the moment should be declared to the injector at bootstrap and have defined dependencies declared also.

The only downside is extensions, which would need to declare themselves to the injector at bootstrap time; So a little re-jigging of the extension loading pattern would be needed (which I am already playing with from a composer autoload standpoint).

Can those interested have a read of the info for this injector? It's a little complex, but that's the point.

- [ ] Make a list of classes and dependencies (I have an old one somewhere)
- [ ] Branch the codebase
- [ ] Make the classes use dependency injection
- [ ] Alter the extension loading pattern to allow dependency injection
- [ ] Alter the bootstrap process to defined dependencies
- [ ] Test it (fingers crossed)

**Comment by michael-e** [comment]
Excuse me if I ask stupid questions, but **why** do you want to remove static classes?

**Comment by designermonkey** [comment]
Because it's a very brittle design pattern that is completely untestable.

**Comment by michael-e** [comment]
Ah, thanks.

**Comment by nitriques** [comment]
> Because it's a very brittle design pattern that is completely untestable

What a convincing argument ;) I totally agree with it. I would keep the static instance tho.

**Comment by designermonkey** [comment]
>I would keep the static instance tho.

Why?

**Comment by nitriques** [comment]
Convenience: Not having to find "Globals". Singleton can be used also.
Performance: Not having to create new object that would be the same thing as others already present in the database.

You could argue that we could use a dictionary or pool of object for those. You would be also right. But I am pretty sure that even that solution would have a static method to get to those objects, right ?

But still, removing static method and using more delegation (broking big classes with too much responsibilities) will be really better and easier to test, and I am all for it.

**Comment by jensscherbl** [comment]
@nitriques http://www.phptherightway.com/#dependency_injection

**Comment by nitriques** [comment]
@jensscherbl Interesting. Any part I should read more carefully ? :)

**Comment by jensscherbl** [comment]
> Any part I should read more carefully ?

Not really, just thought you are new to the concept, based on your last comment. No offence, still wrapping my head around this as well... ;)

**Comment by nitriques** [comment]
No problem, I asked that question because I was curious. I've done a lot of IoC in Java and C# and the article you mention kinda uses the sames patterns/tricks.

**Comment by jensscherbl** [comment]
> I've done a lot of IoC in Java and C# and the article you mention kinda uses the sames patterns/tricks.

Yeah, it's the same pattern.
